Add driver for DS2482-800 bus driver

casparfriedrich opened the pull request.

casparfriedrich committed:
drivers: ds248x: Fix naming

ds248x no longer applies to all drivers. Therefore the naming indicates
compatibility with DS2482 and DS2484 drivers.

Also
- Fix some code formatting

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

casparfriedrich committed:
drivers: w1: Let driver explicitly implement the bus lock functionality

This path can be used as a runtime initialization in a select/deselect
manner.

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

amirfuhrmann commented on drivers/w1/w1_ds2482_800_channel.c:
```diff
@@ -0,0 +1,180 @@
+
+#define DS2482_CHANNEL_INIT(inst)                                                                  \
+	static const struct ds2482_channel_config inst_##inst##_config = {                         \
+		.w1_config.slave_count = W1_INST_SLAVE_COUNT(inst),                                \
```
I think there is something more to consider . how is W1_INST_SLAVE_COUNT really defined ? -- i see in w1.h it is some static derivative of device tree. I need it to be independent from device tree (and large). as I have up to about 128 devices and they are randomly distributed on the 8 w1 ports of the 2482-800. The w1_net layer seems to use this in certain paths . How can I meet my objective ?

casparfriedrich committed:
samples: w1: Update scanner example for ds2482-800 bus driver

The scanner example for 1-wire now supports the DS2482-800 multi channel
bus master.

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

str4t0m reviewed (changes requested):
Hi, thanks for this addition. I have a couple of comments.

dp7hgh7 commented on dts/bindings/w1/maxim,ds2482-800.yaml:
```diff
@@ -0,0 +1,11 @@
+include: [i2c-device.yaml]
+
+on-bus: i2c
+bus: ds2482-800
```
I probably don't get this... This should ensure that the channel driver can only be uses as a node to the parent driver

str4t0m replied:
Indeed it should be fine to use this pattern here. You won't need to change it.

dp7hgh7 commented on drivers/w1/w1_ds2482_800.c:
```diff
@@ -0,0 +1,72 @@
+	return 0;
+}
+
+static const struct w1_driver_api api = {
+	.change_bus_lock = ds2482_change_bus_lock_impl,
+};
```
Good idea. Done.

dp7hgh7 commented on include/zephyr/drivers/w1.h:
```diff
@@ -237,7 +237,7 @@ static inline int z_impl_w1_read_byte(const struct device *dev)
 	const struct w1_driver_api *api = (const struct w1_driver_api *)dev->api;
 
-	return api->read_byte(dev);
+	return api->read_byte ? api->read_byte(dev) : -ENOSYS;
```
So what's your suggestion? Drop this mechanism altogether? I think its='s generally a good idea to return something meaningful instead of crashing, but since the parent driver no longer uses the 1-Wire API this could be dropped.

casparfriedrich committed:
drivers: ds2482-800: Add driver

This adds a driver for the DS2482-800 1-wire multi channel bus driver.
The driver uses a split architecture in order to share a common lock
among all configured channels of a single IC.

Signed-off-by: Caspar Friedrich <c.s.w.friedrich@gmail.com>

str4t0m reviewed (commented):
Looks mostly good, only a few nits.

dp7hgh7 commented on include/zephyr/drivers/w1.h:
```diff
@@ -111,6 +113,12 @@ __syscall int w1_change_bus_lock(const struct device *dev, bool lock);
 {
+	const struct w1_driver_api *api = dev->api;
+
+	if (api->change_bus_lock) {
+		return api->change_bus_lock(dev, lock);
+	}
```
done

dp7hgh7 commented:
Rebased
